src/lib/mina_net2/libp2p_stream.mli

```diff
@@ -0,0 +1,43 @@
+open Async_kernel
+open Network_peer
+
+type participant = Us | Them [@@deriving equal]
+
+type t
+
+val id : t -> Libp2p_ipc.stream_id
+
+(** Open a new stream to a remote peer. *)
+val open_ :
+     logger:Logger.t
+  -> helper:Libp2p_helper.t
+  -> protocol:string
+  -> peer_id:Libp2p_ipc.peer_id
+  -> release_stream:(Libp2p_ipc.stream_id -> unit)
+  -> t Deferred.Or_error.t
+
+(** Create a new stream type from an existing stream. *)
+val create_from_existing :
+     logger:Logger.t
+  -> helper:Libp2p_helper.t
+  -> stream_id:Libp2p_ipc.stream_id
+  -> protocol:string
+  -> peer:Peer.t
+  -> release_stream:(Libp2p_ipc.stream_id -> unit)
+  -> t
+
+val protocol : t -> string
+
+val remote_peer : t -> Peer.t
+
+val pipes : t -> string Pipe.Reader.t * string Pipe.Writer.t
+
+val data_received : t -> string -> unit
+
+val reset : helper:Libp2p_helper.t -> t -> unit Deferred.Or_error.t
+
+val stream_closed :
+     logger:Logger.t
+  -> who_closed:participant
+  -> t
+  -> [ `Stream_should_be_released of bool ]
```

Comment:
In either case for this function's result, we are telling the caller whether or not to garbage collect the stream from any of its data structures so that we don't leak memory once the streams are no longer in use. I'm not sure that this differentiation makes sense to the caller here.